src/main/java/build/buildfarm/worker/Worker.java

```diff
@@ -205,9 +207,20 @@ private boolean linkInputs(Path execDir, Digest inputRoot, Map<Digest, Directory
     for (FileNode fileNode : directory.getFilesList()) {
       Digest digest = fileNode.getDigest();
       Path fileCachePath = cacheDir.resolve(String.format("%s_%d", digest.getHash(), digest.getSizeBytes()));
-      OutputStream outputStream = Files.newOutputStream(fileCachePath);
-      instance.getBlob(digest).writeTo(outputStream);
-      outputStream.close();
+      boolean createFile = !Files.exists(fileCachePath);
+      if (createFile) {
+        OutputStream outputStream = Files.newOutputStream(fileCachePath);
+        instance.getBlob(digest).writeTo(outputStream);
+        outputStream.close();
+      }
+      if (createFile || Files.isExecutable(fileCachePath) != fileNode.getIsExecutable()) {
+        ImmutableSet.Builder<PosixFilePermission> perms = new ImmutableSet.Builder<PosixFilePermission>()
+          .add(PosixFilePermission.OWNER_READ);
+        if (fileNode.getIsExecutable()) {
```

Comment:
If you're sharing the file cache between multiple executions, then you could have a problem with a concurrent execution changing the permissions if there are actions that have the same file but with executable on/off. Is that a concern? Otherwise, I'd like to see it documented somewhere that the file cache is exclusive.